ksmt-core/src/main/kotlin/org/ksmt/decl/KDecl.kt

```diff
@@ -2,19 +2,69 @@ package org.ksmt.decl
 
 import org.ksmt.KAst
 import org.ksmt.KContext
+import org.ksmt.cache.hash
 import org.ksmt.expr.KApp
 import org.ksmt.expr.KExpr
 import org.ksmt.sort.KSort
 
 abstract class KDecl<T : KSort>(
     ctx: KContext,
     val name: String,
-    val sort: T
+    val sort: T,
+    val argSorts: List<KSort>
 ) : KAst(ctx) {
     abstract fun apply(args: List<KExpr<*>>): KApp<T, *>
     abstract fun <R> accept(visitor: KDeclVisitor<R>): R
 
-    //  Contexts guarantee that any two equivalent declarations will be the same kotlin object
-    override fun hashCode(): Int = System.identityHashCode(this)
-    override fun equals(other: Any?): Boolean = this === other
+    override fun print(builder: StringBuilder): Unit = with(builder) {
+        append('(')
+        append(name)
+
+        if (this@KDecl is KParameterizedFuncDecl) {
+            append(parameters.joinToString(separator = " ", prefix = " [", postfix = "]"))
+        }
+
+        append(" (")
+
+        for ((i, sort) in argSorts.withIndex()) {
+            if (i > 0) {
+                append(" ")
+            }
+
+            sort.print(this)
+        }
```

Comment:
Why not `joinToString` with a lambda?

Reply:
`sort.print` modifies provided `StringBuilder` and doesn't return a string.

Reply:
Candidate for core extension function if code like this happens more than once